src/java/org/apache/cassandra/service/accord/AccordCommand.java

```diff
@@ -0,0 +1,788 @@
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.ObjectSizes;
+import org.apache.cassandra.utils.concurrent.AsyncPromise;
+import org.apache.cassandra.utils.concurrent.Future;
+
+import static org.apache.cassandra.service.accord.AccordState.WriteOnly.applyMapChanges;
+import static org.apache.cassandra.service.accord.AccordState.WriteOnly.applySetChanges;
+
+public class AccordCommand extends Command implements AccordState<TxnId>
+{
+    private static final Logger logger = LoggerFactory.getLogger(AccordCommand.class);
+
+    private static final AtomicInteger INSTANCE_COUNTER = new AtomicInteger(0);
+
+    private static final long EMPTY_SIZE = ObjectSizes.measure(new AccordCommand(null, null));
+
+    public static class WriteOnly extends AccordCommand implements AccordState.WriteOnly<TxnId, AccordCommand>
+    {
+        private Future<?> future = null;
+
+        public WriteOnly(AccordCommandStore commandStore, TxnId txnId)
+        {
+            super(commandStore, txnId);
+        }
+
+        @Override
+        public void future(Future<?> future)
+        {
+            Preconditions.checkArgument(this.future == null);
+            this.future = future;
+        }
+
+        @Override
+        public Future<?> future()
+        {
+            return future;
+        }
+
+        @Override
+        public void applyChanges(AccordCommand instance)
+        {
+            applyMapChanges(this, instance, cmd -> cmd.waitingOnCommit);
+            applyMapChanges(this, instance, cmd -> cmd.waitingOnApply);
+            applySetChanges(this, instance, cmd -> cmd.blockingCommitOn);
+            applySetChanges(this, instance, cmd -> cmd.blockingApplyOn);
+        }
+    }
+
+    public static class ReadOnly extends AccordCommand implements AccordState.ReadOnly<TxnId, AccordCommand>
+    {
+        public ReadOnly(AccordCommandStore commandStore, TxnId txnId)
+        {
+            super(commandStore, txnId);
+        }
+
+        @Override
+        boolean isReadOnly()
+        {
+            return true;
+        }
+    }
+
+    private final AccordCommandStore commandStore;
+    private final TxnId txnId;
+    private final int instanceCount = INSTANCE_COUNTER.getAndIncrement();
+    public final StoredValue<Key> homeKey;
+    public final StoredValue<Key> progressKey;
+    public final StoredValue<Txn> txn;
+    public final StoredValue<Ballot> promised;
+    public final StoredValue<Ballot> accepted;
+    public final StoredValue<Timestamp> executeAt;
+    public final StoredValue<Deps> deps;
+    public final StoredValue<Writes> writes;
+    public final StoredValue<Result> result;
+
+    public final StoredValue.HistoryPreserving<Status> status;
+    public final StoredBoolean isGloballyPersistent;
+
+    public final StoredNavigableMap<TxnId, ByteBuffer> waitingOnCommit;
+    public final StoredNavigableMap<TxnId, ByteBuffer> waitingOnApply;
```

Comment:
FYI, this should be sorted by `Timestamp` (executeAt). But I will address it in my follow-up patch.